Animation Editor - First Version

mattigruener opened the pull request:
Hey guys,

I cleaned up what was a stack of ugly commits and I think it should be fairly clean now. This PR contains both John's work on the Animation API and my work on the Animation Editor that makes use of the former.

The intention here is to get out a first version in order to get some feedback. I believe it comes with the features needed for a reasonable first version: visualization of curves/keys and insertion/deletion/moving of keyframes.

There are a few things that I wanted to mention in terms of limitations, gotchas and future work. In no particular order:

- We had a discussion recently about `MMB` dragging in the Node Graph. I wanted to bring up that in order to pan the camera in the Animation Editor users will have to press `alt`, which is different to what the Node Graph offers. A simple `MMB` drag in the Animation Editor is used to move selected keyframes. This is probably easy enough to get used to.

- We should probably work on undo support next? John, do you think the lack of it prevents us from getting this in peoples' hands?

- John, I'm doing something in the `AnimationPathFilter` in `AnimationEditor.py` that feels like an abuse. I'm emitting the `changedSignal()` for a bunch of changes to trigger an update in the `PathListingWidget`. I'm sure you'll have a better idea regarding how to get the Widget to update properly?

- Similar to what you've done in the `AuxiliaryConnections` refactoring, John, I'm thinking that I should get rid of the extra `CurveGadget`. That would make it easier for us if we want to select all keys by clicking on a curve, for example, - or if we want to show a tooltip for a key that shows it's time and value data. That could be a change at a later stage, though.

When talking to Michael we said I'd put a comment on this PR regarding hotkeys that need to be documented somewhere. Here are a few things users should know about:
- In addition to normal `alt`+`RMB` zooming, users can use `shift`+`alt`+`RMB` to zoom each axes individually
- Selected keys are moved using the `MMB` (hold `shift` to limit movement to one axes).
- `f` can be pressed to frame selected keys or visible curves
- `i` can be pressed to insert a keyframe into all editable curves
- `delete` can be pressed to delete selected keys

John, would you take a look at the code? I know you don't want to stomp over what I've done here, but it'd be great if you could point out what could be improved :)

Thanks,

Matti.

johnhaddon commented:
Great! Thanks Matti! Before I get started testing and reviewing, I just wanted to reply to this one question of yours...

> We should probably work on undo support next? John, do you think the lack of it prevents us from getting this in peoples' hands?

I do think we need everything to be undoable before we can release this, yes. Undo is fundamental to a decent user experience...

Cheers...
John

mattigruener commented:
I believe you had some thoughts on undo already - in particular how to merge continuous changes in a drag. All of that is probably more on the Animation API side of things? Shall we offline that discussion and see how much of that I can do and how much of your time we might still need?

Thanks for looking through it :)

Cheerio

johnhaddon commented:
> I believe you had some thoughts on undo already - in particular how to merge continuous changes in a drag. All of that is probably more on the Animation API side of things? Shall we offline that discussion and see how much of that I can do and how much of your time we might still need?

For a first version I think it might be acceptable to just use the API as it is, and treat the stuff I was talking about as a later optimisation. Is that correct, or is something actually blocking the implementation of undo?

mattigruener commented:
> Is that correct, or is something actually blocking the implementation of undo?

No, if we're ok with that potentially not being as fast as it could be, then I don't think there's a problem. I took a quick stab at it in fa9288d. I think I've stumbled over this before - it might be nice to have something along the lines of `closeLastMergeGroup()` for the `UndoScope` at some point? What I've done now, feels a little bit hacky but gets the job done, I guess :)

johnhaddon commented:
Hi Matti,

I thought I'd put this through its paces before looking at the code, so I could come at it as a naive user. Based on that experience I think we're super close, but there are some simple refinements that are probably worth making before we roll out a first version...

- I figured out fairly easily how the selection/move stuff worked, but found the following a bit confusing :
  - The highlighting of the curve on hover. This made me think I could click on it to select it, but that didn't seem to work. Coupled with the need to select the curve to get at the control points, this made it a bit tricky to figure out how to get into an editing state.
  - Clicking on control points didn't seem to select them - I always had to drag select.
  - I didn't find MMB to move to be all that intuitive - elsewhere in Gaffer MMB is for a secondary operation. Can we use Left-drag for the move? This would also allow a single-click-drag operation to both select a key and move it. And it would free up MMB for the viewport motion again.
   - The blue of selected keys doesn't stand out against the white of unselected keys.
   - I couldn't for the life of me figure out how to insert a key. The thing that seemed most obvious to try was Ctrl+click, given how Dots are made, but I tried all sorts of other combinations too.
- The colour of the vertical "current frame" bar is hard to distinguish from the colour of the X curves. I managed to open my first graph in a state that convinced me I had some weird off-to-infinity drawing error, because it all looked like the one curve.
- Framing seemed to behave a bit oddly at times, including a lot of negative frames. Perhaps we can use the ScriptNode framerange as a hint there?
- Hitting `Backspace` doesn't delete keys but instead falls through to the main menu and deletes the selected node. I suspect this is just a case of needing to bind to `Backspace` as well as `Delete`.
- Renaming a node resets the expansion/selection in the AnimationEditor.
- Misc aesthetic stuff
   - The grid lines seemed a tad heavy to my eye.
   - The dark block background on the axes seemed even heavier. Perhaps they could use the same grey as inactive tabs?
   - A little frame/border around the whole thing would help a bit I think.

Any thoughts?
Cheers..
John

johnhaddon commented:
> i can be pressed to insert a keyframe into all editable curves

Ha! Just realised I skipped the documentation on that one! I still couldn't make it work though for some reason, and think Ctrl+Click would be nice...

mattigruener commented on src/GafferUI/AnimationGadget.cpp:
```diff
@@ -468,6 +485,14 @@ void AnimationGadget::removeKeyframes()
+	}
+
+	// There have to be editable plugs if there are selected keys
+	UndoScope undoEnabled( m_curvePlugsEditable[0]->ancestor<ScriptNode>(), UndoScope::Enabled, m_dragUndoToggle ? "animationEditorMoveA" : "animationEditorMoveB" );
```
Ah, I didn't consider the case of having multiple Animation Editors open. Thanks - I fixed it in ba91a45 following your example in the `TransformTool`.

mattigruener commented:
Hey John,

thanks for giving it a go :) Everything I'm not commenting on I'll just try to address in the next batch of code I put up.

> The highlighting of the curve on hover. This made me think I could click on it to select it, but that didn't seem to work. Coupled with the need to select the curve to get at the control points, this made it a bit tricky to figure out how to get into an editing state.

Agreed.  Some of the other things you brought up also suggest that I should maybe look into removing the `CurveGadget` now and replace it with something managed by the `AnimationGadget`. At the same time, I could make the curves selectable. Does that sound good?

> Clicking on control points didn't seem to select them - I always had to drag select.

True, will fix that and this alone requires picking to be managed somewhat by the AnimationGadget, I guess.

> I didn't find MMB to move to be all that intuitive - elsewhere in Gaffer MMB is for a secondary operation. Can we use Left-drag for the move? This would also allow a single-click-drag operation to both select a key and move it. And it would free up MMB for the viewport motion again.

I just double-checked what a big software package with 4 letters in the name is doing. 

Left-dragging anywhere on the background creates a selection box, left-dragging on a _selected_ key moves that key - but left-dragging on an _unselected_ key again just creates a selection box. Also, the `MMB` drag works in the same way the AnimationEditor handles it currently - simple `MMB` drag moves selected keys if any, `alt + MMB` pans. 
I'll look into getting the left-dragging working the same, except that you can select-drag with a single click-drag. Regarding `MMB`-dragging, maybe what people know from Gaffer trumps what people know from other packages (even though they are the de-facto industry standard for animation) and we should remove the `MMB` draggging? I think I implemented it that way because I remembered maya supporting it - maybe I was always too lazy to precisely click on a key and only knew this way. Maybe other people are the same?

> The blue of selected keys doesn't stand out against the white of unselected keys.

Black keys?

> I couldn't for the life of me figure out how to insert a key. The thing that seemed most obvious to try was Ctrl+click, given how Dots are made, but I tried all sorts of other combinations too.

Haha, sorry - I thought `i` might be somewhat discoverable, and Maya apparently did it that way until 2017 when that hotkey was changed. I like the ctrl-click idea - my only concern is that it's gonna be useful in some cases to insert a key into *all* editable curves at the current frame, and without having to aim for the current frame with the mouse. That's why I thought a hotkey is useful. Maybe we just support both? Would you expect a keyframe preview like we're rendering for the dot inserting?

> The colour of the vertical "current frame" bar is hard to distinguish from the colour of the X curves. I managed to open my first graph in a state that convinced me I had some weird off-to-infinity drawing error, because it all looked like the one curve.

You're right, - yellow line? That shouldn't clash with the default colors for XYZ and we can make it a bit darker to not be too close to white curves.

> Framing seemed to behave a bit oddly at times, including a lot of negative frames. Perhaps we can use the ScriptNode framerange as a hint there?

Mh, I think I haven't seen this. Maybe because my Gaffer always starts at 1001... You're suggesting to intersect the computed range to be framed with the current framerange? Did I get that correctly? 

Cheerio :)

johnhaddon commented:
> Some of the other things you brought up also suggest that I should maybe look into removing the CurveGadget now and replace it with something managed by the AnimationGadget. At the same time, I could make the curves selectable. Does that sound good?

Honestly, I'm not sure on the CurveGadget question. For the "click on curve to select" thing, might it actually be easier to keep it? As long as CurveGadget remains private, I don't think this is hugely important - whatever works.

> I'll look into getting the left-dragging working the same, except that you can select-drag with a single click-drag.

Great!

> Regarding MMB-dragging, maybe what people know from Gaffer trumps what people know from other packages (even though they are the de-facto industry standard for animation) and we should remove the MMB draggging? I think I implemented it that way because I remembered maya supporting it - maybe I was always too lazy to precisely click on a key and only knew this way. Maybe other people are the same?

Worth surveying the floor I think - as long as left-drag works too, I don't have a strong opinion on middle drag. I've got into the habit of always holding Alt to move the camera, even when it's not needed.

> > The blue of selected keys doesn't stand out against the white of unselected keys.
> Black keys?

Will that stand out enough against the grid lines? 

> I like the ctrl-click idea - my only concern is that it's gonna be useful in some cases to insert a key into all editable curves at the current frame, and without having to aim for the current frame with the mouse. That's why I thought a hotkey is useful. Maybe we just support both?

Both sounds good to me. This seems slightly lower priority to me than the other stuff I mentioned - it could come later if needs be.

> Would you expect a keyframe preview like we're rendering for the dot inserting?

That would be a nice touch for sure.

> yellow line? That shouldn't clash with the default colors for XYZ and we can make it a bit darker to not be too close to white curves.

Yeah, let's give that a go. If it looks good, we should probably update the Timeline to use the same style so that they're intuitively linked. We can do the Timeline separately though...

> Mh, I think I haven't seen this. Maybe because my Gaffer always starts at 1001... You're suggesting to intersect the computed range to be framed with the current framerange? Did I get that correctly?

Here's some steps to reproduce :

- Have a Script frameRange of 1-100
- Create a Plane
- Using the NodeEditor create the following keys for the `dimensions.x` plug :
    - Frame 1 : 1
    - Frame 100 : 10
- Open the AnimationEditor, and expand all on the left
- Hit 'F' to frame

With that, I get a range of -100 to 250 framed in the X axis, even though the keys only occupy the 1 to 100 range. Ideally I think I would expect to see only 1-100 framed, but I definitely don't expect to see negative frames. I don't think we want to intersect with the Script frame range though, because that would omit keys that might live outside it. I think the issue might be that the framing need to use "decoupled axis zooming" - it looks like I'm getting a vertical fit and then getting whatever horizontal fit is dictated by that, rather than getting separate fits in each direction.

Oh, one other thing I think I forgot to mention : we should add shiny new AnimationEditors into all the standard layouts. Maybe in the same panel as the NodeGraph, since they have similar dimensions, and you're likely to want to be looking at the results of the animation in the Viewer?

mattigruener commented:
For the record - one of the tests is failing:

```
FAIL: testScaling (GafferDispatchTest.LocalDispatcherTest.LocalDispatcherTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/GafferHQ/gaffer/install/gaffer-0.47.0.0-linux/python/GafferDispatchTest/LocalDispatcherTest.py", line 784, in testScaling
    self.assertLess( time.clock() - t, timeLimit )
AssertionError: 20.031062 not less than 6
```

I don't think that's me. It's just one of the four setups and maybe just needs a retry, but I thought I'd save the error message for future reference.

mattigruener commented:
Quoting myself here:

> You can now click on a non-editable curve and make it editable that way. I think it should respect the shift key, though, and replace the list of editable curves and only add to it when shift is pressed.

I just squashed in changes that should make that work as I described.

mattigruener commented:
I've addressed your comment about integrating the Animation Editor into the default layouts. @donboie had an improvement that he did on his spreadsheet branch ( 3156d8f  ). We thought it makes things a little bit more maintainable - what do you think? I made it part of this PR as this one is probably gonna get merged before his spreadsheet PR.

mattigruener commented:
I've added a dedicated class for the slider used in the timeline in order to override how the position indication is drawn. We should probably also look into whether we can make it snap to frames. Are you ok with the approach I've taken here? I've only changed the color for now.

johnhaddon commented on python/GafferUI/TimelineSlider.py:
```diff
@@ -0,0 +1,63 @@
+from Qt import QtWidgets
+
+# \todo: documentation
+class TimelineSlider( GafferUI.NumericSlider ) :
```
Let's just keep this privately in Timeline.py.

mattigruener replied:
Done, I've also changed the drawing. It's a glorious yellow line now. We've always had a problem with snapping when dragging the slider backwards - it feels like that's a bit more noticeable now?

johnhaddon commented on include/GafferUI/AnimationGadget.h:
```diff
@@ -0,0 +1,189 @@
+
+	void setVisiblePlugs( const std::vector<Gaffer::Plug *> &plugs );
+	void setEditablePlugs( const std::vector<Gaffer::Plug *> &plugs );
+	void setFrame( float frame );
```
Although we don't currently expose it to users, the API is such that different editors can use different contexts (see `Editor.setContext()`). Really the AnimationGadget needs a setContext/getContext pair rather than a frame for this to continue to work. Internally we shouldn't be manipulating the ScriptNode context directly...

mattigruener replied:
Addressed this one, too. I hope it's better now.

johnhaddon commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -125,6 +125,11 @@ class GAFFERUI_API ViewportGadget : public Gadget
 
+		/// When variable aspect zoom is enabled, the two axis can be scaled
+		/// independently when performing a 2D zoom.
+		void setVariableAspectZoom( bool variableAspectZoom );
```
@danieldresser, any comments as to how this will interact with the work you're doing?

johnhaddon commented on src/GafferUIModule/AnimationGadgetBinding.cpp:
```diff
@@ -70,14 +75,35 @@ void setEditablePlugs( AnimationGadget &animationCurveGadget, object plugs )
+		}
+		catch( const error_already_set &e )
+		{
+			PyErr_PrintEx( 0 ); // clears the error status
```
Just spotted this - it should be `translatePythonException()`, but if we use StandardSets like I suggested, I don't think we'll need this signal anyway...

mattigruener commented on include/GafferUI/AnimationGadget.h:
```diff
@@ -0,0 +1,154 @@
+
+/// Aliases that define the intended use of each
+/// Gadget::Layer by the AnimationGadget components.
+namespace AnimationLayer
```
Done.

mattigruener commented on include/GafferUI/AnimationGadget.h:
```diff
@@ -0,0 +1,154 @@
+	constexpr Gadget::Layer Overlay = Gadget::Layer::Front;
+};
+
+struct KeyPtrLessThan
```
That was gone already.

mattigruener commented on include/GafferUI/AnimationGadget.h:
```diff
@@ -0,0 +1,154 @@
+class GAFFERUI_API AnimationGadget : public Gadget
+{
+
+public :
```
Should be better now.

mattigruener commented on include/GafferUI/AnimationGadget.h:
```diff
@@ -0,0 +1,189 @@
+
+	IE_CORE_DECLARERUNTIMETYPEDEXTENSION( GafferUI::AnimationGadget, AnimationGadgetTypeId, Gadget );
+
+	void setVisiblePlugs( const std::vector<Gaffer::Plug *> &plugs );
```
I've implemented that. The cpp side of things got a little messier because of added casts, but that's probably ok. Thanks for the suggestion!

mattigruener commented on src/GafferUIModule/AnimationGadgetBinding.cpp:
```diff
@@ -0,0 +1,109 @@
+#include "GafferUIBindings/GadgetBinding.h"
+#include "GafferUI/AnimationGadget.h"
+
+#include "AnimationGadgetBinding.h"
```
Fixed.

mattigruener commented on startup/gui/layouts.py:
```diff
@@ -50,9 +52,81 @@
+	)
+
+def getSource(fn):
+	a = inspect.getsource(fn).split("\n")
```
I've switched that over to a triple quoted string. That change is in a separate commit at the moment - maybe we just squash all three of those commits in the end. I was rushing a fair bit with all of this and skipped separating the general formatting change from adding the AnimationEditor.

johnhaddon commented:
> > I also found a repeatable crash
> Maybe I did it wrong, but I couldn't reproduce that. If this still happens for you after the restructuring I've done today, I'll take a look at it as soon as I can.

Definitely still happening for me, every time I've tried so far. Maybe it's a MacOS/Linux or libstdc++/libc++ thing, and that's why you're not seeing it? I'll check again on Linux when I'm in the office tomorrow. If it only reproduces on MacOS I'll take a look at fixing it myself...

mattigruener commented:
Hey John,

I still couldn't replicate the crash you're seeing, but in order to ease back in after some time off, I've addressed some of the things we had identified as nice to have at some point:

- Dragging the line (and the label) indicating the current frame should work now
- The `Control`-click feature for adding keys has a preview now (pretty much like the dot adding we do for `StandardConnectionGadgets`).

Cheerio,

Matti.

johnhaddon commented:
Welcome back!

> I still couldn't replicate the crash you're seeing

I couldn't reproduce it on Linux either but it still happens on mac every time. [Here's a fix](https://github.com/johnhaddon/gaffer/commit/95af3a37f8de30f3d4b9e5b83e62e8553abfbe55). I also needed to fix the [import of TimelineSlider](https://github.com/johnhaddon/gaffer/commit/d2cbd50e63f10c60fb88eff5ff0a3e36cd219f01) before I could run it successfully.

> I've addressed some of the things we had identified as nice to have at some point

Nice! Thanks!

Could you add my fixes and squash the various review changes back into the relevant initial commits?

mattigruener commented:
Hey John,

> Welcome back!

Good to be back! I'm excited to get this one out there :)

Thanks for the fix - I've squashed it into the initial commit to eliminate every trace of my stupidity. 

I've cleaned up the history regarding the other commits, as well. Also, since Don has reverted his changes to how the default layouts are represented in his spreadsheet branch, I've done the same to make it easier to eventually merge the two.

Thanks :)

Matti.

mattigruener commented:
Rebased onto latest master to pick up `SceneHierarchy -> HierarchyView` name change.

johnhaddon commented:
Thanks Matti! I've taken a very quick cleanup pass over this in preparation for merging. Could you take a quick look at the following commits and let me know if I've mucked anything up?

https://github.com/johnhaddon/gaffer/commits/mattigruener-curveEditorIE

mattigruener commented:
I've added a commit that allows showing curves for nodes nested in nodes. Not sure if that ever worked before, but a small change in one of the cherry-picked commits would've broken it anyway. I've reverted that change (look at all children, not just plugs) and cleaned up/fixed some other code to make things more readable.

johnhaddon commented:
Thanks Matti, I've merged manually to fix the conflict with `startup/gui/layouts.py`. Daniel's comments make sense to me, but should we wait for a first round of feedback from a few more users before deciding on next steps?

johnhaddon commented:
> Daniel also sent me a crash log. He said he used the AnimationEditor for a few hours and it happened once "while attempting to use shift to offset several selected curves horizontally in time". 

I should explain my thinking on this one. I've taken a quick look at the code and don't see anything obviously wrong. And since it took hours to trigger, I don't think we're going to get to the bottom of it by leaving the feature rotting in a PR. Call me reckless, but I think the only pragmatic thing to do is merge and hope that leads to us finding better steps to reproduce...

mattigruener commented:
> Daniel's comments make sense to me, but should we wait for a first round of feedback from a few more users before deciding on next steps?

Yes, sounds good. Let's see how people use and abuse it.

> I think the only pragmatic thing to do is merge and hope that leads to us finding better steps to reproduce

Agreed. 

Thanks for merging and your help with this one!
